include/parallel/parallel_implementation.h

```diff
@@ -3302,6 +3302,90 @@ inline void Communicator::allgather(std::vector<T> & r,
 
 
 template <typename T>
+inline void Communicator::allgather(std::vector<std::basic_string<T> > & r,
+                                    const bool identical_buffer_sizes) const
+{
+  if (this->size() < 2)
+    return;
+
+  LOG_SCOPE("allgather()", "Parallel");
+
+  if (identical_buffer_sizes)
+    {
+      libmesh_assert(this->verify(r.size()));
+
+      // identical_buffer_sizes doesn't buy us much since we have to
+      // communicate the lengths of strings within each buffer anyway
+      if (r.empty())
+        return;
+    }
+
+  // Concatenate the input buffer into a send buffer, and keep track
+  // of input string lengths
+  std::vector<int> mystrlengths (r.size());
+  std::vector<T> concat_src;
+
+  int myconcatsize = 0;
+  for (unsigned int i=0; i != r.size(); ++i)
+    {
+      int stringlen = cast_int<int>(r[i].size());
+      mystrlengths[i] = stringlen;
+      myconcatsize += stringlen;
+    }
+  concat_src.reserve(myconcatsize);
+  for (unsigned int i=0; i != r.size(); ++i)
+    concat_src.insert
+      (concat_src.end(), r[i].begin(), r[i].end());
+
+  // Get the string lengths from all other processors
+  std::vector<int> strlengths = mystrlengths;
+  this->allgather(strlengths, identical_buffer_sizes);
+
+  // We now know how many strings we'll be receiving
+  r.resize(strlengths.size());
+
+  // Get the concatenated data sizes from all other processors
+  std::vector<int> concat_sizes;
+  this->allgather(myconcatsize, concat_sizes);
+
+  // Find the total size of the final concatenated array and
+  // set up the displacement offsets for each processor.
+  std::vector<int> displacements(this->size(), 0);
+  unsigned int globalsize = 0;
+  for (unsigned int i=0; i != this->size(); ++i)
+    {
+      displacements[i] = globalsize;
+      globalsize += concat_sizes[i];
+    }
+
+  // Check for quick return
+  if (globalsize == 0)
+    return;
+
+  StandardType<T> send_type(&concat_src[0]);
```

Comment:
Have you tested the stupid case where the vector of strings is initially empty on all processors? I think `concat_src` might be empty then (?) and indexing into the zeroth entry could fail.

Reply:
Damn, you're right.  The stupid "empty on all processors" case should only work if the user set identical_buffer_sizes.  The "empty on one processor" case (which is not stupid, I'm pretty sure my use case will actually hit it on not-too-crazy processor counts) won't work at all.  I'll put it in a fixup since I need to interactive-rebase anyway to fix that commit message.

Reply:
Thanks for catching that!